### ISSUE TITLE
[Question] When should I use FusedEmbeddingBagCollection over EmbeddingBagCollection?

### ISSUE THREAD

**Issue by JacoCheung** [meta-pytorch/torchrec#1741]
There is a [benchmark](https://github.com/pytorch/torchrec/tree/main/benchmarks) under torchrec repo which compares **EmbeddingBagCollection** backed by nn.EmbeddingBag and **FusedEmbeddingBagCollection** backed by FBGEMM. However, AFAIK, there should be a non-fused **EmbeddingBagCollection** backed by FBGEMM

To my knowledge, there should be: 

- **EmbeddingBagCollection** backed by nn.EmbeddingBag
- **EmbeddingBagCollection** backed by FBGEMM (The underlying implementation should ShardedEmbeddingBagCollection)
- **FusedEmbeddingBagCollection** backed by FBGEMM (It also can be sharded, but the sharder is not present in the torchrec [benchmark](https://github.com/pytorch/torchrec/tree/main/benchmarks).)

I have few questions: 

1. To leverage FBGEMM kernel, the EmbeddingBagCollection must be used together with sharder, right?
2. In [DLRM repo](https://github.com/facebookresearch/dlrm/blob/main/torchrec_dlrm/dlrm_main.py#L595-L628), the EmbeddingBagCollection rather than FusedEmbeddingBagCollection is used. So I wonder when should I use FusedEmbeddingBagCollection.
3. I'm curious if there is any performance result of non-fused EmbeddingBagCollection backed by FBGEMM comparing to FuedEmbeddingBagCollection. Especially for multi-GPU cases.

Thanks!

**Comment by henrylhtsang** [comment]
I think we are deprecating FusedEmbeddingBagCollection. EmbeddingBagCollection is the preferred api going forward.

Also, when you used the EmbeddingComputeKernel.FUSED kernel (as opposed to EmbeddingComputeKernel.DENSE), tables will be fused together and are backed by FBGEMM kernels.

**Comment by JacoCheung** [comment]
Thanks for your reply @henrylhtsang. I assume `Fused*Sharder` will be deprecated as well, right?

How does `EmbeddingComputeKernel.FUSED` differ from `EmbeddingComputeKernel.DENSE` ? 

Does `EmbeddingComputeKernel.DENSE` mean the underlying implementation is  [FBGEMM::DenseTableBatchedEmbeddingBagsCodegen](https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/fbgemm_gpu/split_table_batched_embeddings_ops_training.py#L2021) which returns a dense gradient matrix in the backward process, while `EmbeddingComputeKernel.FUSED` refers to  [FBGEMM::SplitTableBatchedEmbeddingBagsCodegen](https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/fbgemm_gpu/split_table_batched_embeddings_ops_training.py#L281) which returns sparse gradients? 

Does that mean there's no any fusion in `EmbeddingComputeKernel.DENSE`?

**Comment by JacoCheung** [comment]
Also, should `EmbeddingComputeKernel.DENSE` alwasys be paired with  `ShardingType.DATA_PARALLEL`  (**Dp{Pooled}EmbeddingSharding**)?

**Comment by henrylhtsang** [comment]
@JacoCheung 

Yeah the fused*sharders will be deprecated too.

FUSED means we fuse the optimizer and embedding tables together. So when we call loss.backward(), we automatically update the tables already. Also this means there is no need to have an additional optimizer, since everything is done in the split-TBE.

FUSED is more performant than DENSE, empirically.

I think you can still use DENSE for all kinds of sharding type. For data_parallel, you must use DENSE.

**Comment by JacoCheung** [comment]
Thanks for your explanation @henrylhtsang. I think DENSE means a dense gradients for all embedding rows. 

Can I manage to get the sparse gradients (something like KeyedJaggedTensor) by diabling fuse optimizer? (From #1293 , I think there was a way of doing so. But how can I make it?)

**Comment by henrylhtsang** [comment]
@JacoCheung not sure if I understand the question.

If you use FUSED kernel, you can't get the gradients. You would need DENSE for that.
